### ISSUE TITLE
Issues Fine Tuning Blender Bot's 90M Model

### ISSUE THREAD

**Issue by NomanSaleem4** [facebookresearch/ParlAI#3283]
**Bug description**
Firstly model dictionary does not contain the words from my dataset, secondly even after fine tuning the model on my own data bot does not responds as expected even on the utterances from the training data.

**Reproduction steps**
I install the parlai using following steps:
1- git clone https://github.com/facebookresearch/ParlAI.git
2- cd ParlAI/
3- python setup.py develop

Fine Tuned the model using the below command:
```
python -m parlai.scripts.train_model -t fromfile:parlaiformat --fromfile_datapath thumbcrowd_data.txt -m transformer/generator --multitask-weights 1 --num-epochs 1.0 --beam-block-ngram 3 --beam-context-block-ngram 3 --init-model zoo:tutorial_transformer_generator/model --dict-file zoo:tutorial_transformer_generator/model.dict --embedding-size 512 --n-layers 8 --ffn-size 2048 --dropout 0.1 --n-heads 16 --learn-positional-embeddings True --n-positions 512 --variant xlm --activation gelu --skip-generation True --fp16 True --text-truncate 256 --label-truncate 256 --dict-tokenizer bpe --dict-lower True -lr 1e-06 --optimizer adamax --lr-scheduler reduceonplateau --gradient-clip 0.1 -veps 0.25 --betas 0.9,0.999 --update-freq 1 --attention-dropout 0.0 --relu-dropout 0.0 --skip-generation False -vp 15 -stim 60 -vme 20000 -bs 32 -vmt ppl -vmm min --save-after-valid True --model-file  custom_model/test_train_90M --num_epochs 200  
```

Started the Bot using the below command:
```
python parlai/scripts/interactive.py -t blended_skill_talk -mf custom_model/test_train_90M --beam-min-length 20 --beam-block-ngram 3 --beam-context-block-ngram 3 --beam-size 10 --inference beam
```

My Tiny Dataset:
```
text:hello how are you today?	labels:i'm great thanks! how can i help you?
text:what is thumbcrowd	labels:We built ThumbCrowd to empower enterprises to deliver a new and revolutionary AI experience for their customers.		episode_done:True
text:tell me about thumbcrowd	labels:We built ThumbCrowd to empower enterprises to deliver a new and revolutionary AI experience for their customers.	episode_done:True
text:what are the features of thumbcrowd	labels:Create conversational ai assistants that go beyond basic FAQs	episode_done:True
```

**Expected behavior**
1- Bot should get fine tuned on my dataset and on asking the utterance from the dataset, it should predict the correct label.
2- Words from my dataset should be there in the dictionary

Put a lot of effort understanding the blender bot and resolving this problem, but couldn't resolve it. Any help would be much appreciated. Thanks

**Comment by stephenroller** [comment]
The training command and interactive command look right to me. I will note that you're like to overfit on that data if you're not also multitasking on BST.

No words should be added to the dictionary though. We use a BPE dictionary, so unknown words are broken into sub-com-ponents (see what I did there?)

What is the observed behavior that's wrong that you're reporting?

**Comment by NomanSaleem4** [comment]
Hi @stephenroller thanks for the reply, I didn't get this line ```I will note that you're like to overfit on that data if you're not also multitasking on BST. ```  And for BPE dictionary can you point out the code where i have to check that. 
And for the last point, Actually when i enter the text  ```tell me about thumbcrowd``` bot should respond with ```We built ThumbCrowd to empower enterprises to deliver a new and revolutionary AI experience for their customers.``` but it does not, so fine-tuning does not fulfill my use case here. 
Also for how many epochs should i train the bot or should i train until its get trained having ppl into consideration? Thanks :)

**Comment by stephenroller** [comment]
the `--dict-tokenizer bpe` arg is what is sets the bpe mode. You can look at the .dict file for all the words and subwords represented.

I would expect you would see ppl drop to very close to 1 within a few validations. You're overfitting on a single batch, it's generally super easy to do that. It's so fast that we use it for unit testing...

Can you post the training log? And what the model actually says given the input?

Again, you're going to overfit like this. The model is going to forget how to say everything except these 4 utterances, but it should have no trouble saying these utterances perfectly.

**Comment by NomanSaleem4** [comment]
Just gone through test_train_90M.dict and it does not have words from my custom data. When I had trained the bot on these 4 utterances, to drop the ppl close to 1, it took fews hours on cpu.
For training log I need to retrain the bot, i can share logs a little later.
Also how can I avoid the model from being overfit and to avoid the bot being messy for other utterances? Thanks

**Comment by stephenroller** [comment]
As I said, you're not going to have any words added to the dictionary. They aren't in the dictionary, so "ThumbCrowd" is likely to be represented as something like "thumb" + "crowd".

To avoid overfitting, multitask on blended skill talk. So instead of `-t fromfile:parlaiformat` do `-t blended_skill_talk,fromfile:parlaiformat`.

**Comment by NomanSaleem4** [comment]
oh my bad, Great now I got you. Right, I will retrain the model and will let you now. Thank you so so much :)

**Comment by stephenroller** [comment]
Closing, repoen or file new issues if there are more troubles.
